Ad hoc solution to fix false flags

KimSarabia opened the pull request:
This is an ad hoc solution to reduce the amount of false flags before we can switch from pa11y to axe-core for error reporting.

Results from last run:

```
❯ node runA11yTests.js --config 'a11y.config.js'
'Main_Page' - 0 color contrast violations
'Special%3ASearch' - 0 color contrast violations
'YouTube' - 11 color contrast violations
'Cookie_(informatique)' - 84 color contrast violations
'Wikip%C3%A9dia%3AAccueil_principal' - 11 color contrast violations
'Gabriel_Attal' - 49 color contrast violations
'81st_Golden_Globe_Awards' - 3 color contrast violations
'Saltburn_(film)' - 5 color contrast violations
Total errors across all tests: 163
```

> 'Cookie_(informatique)' - 84 color contrast violations 

is still wrong. Will look into this more tomorrow.

Commit:
Ad hoc solution to fix false flags

jdlrobson commented on runA11yTests.js:
```diff
@@ -17,6 +17,34 @@ function resetReportDir( config ) {
+	const colorContrastInstances = [];
+	let addingToArray = false;
+
+	for ( let i = 0; i < data.issues.length; i++ ) {
```
[Suggestion] I'm wondering if you could use data.issues.filter.map here -t might be more readable?

const colorContrastIssues = data.issues.filter((issue)=> issue.code === 'color-contrast')

jdlrobson commented on runA11yTests.js:
```diff
@@ -17,6 +17,34 @@ function resetReportDir( config ) {
+			// Add the current issue to the array
+			colorContrastInstances.push( issue );
+		} else if ( addingToArray && i > 0 && data.issues[i].code !== 'color-contrast' && data.issues[i - 1].code === 'color-contrast' ) {
+			// Break the loop if a different code value is encountered after adding to the array
```
[Question] Why would we do this? If I'm reading correctly if data.issues looks like:

```
[ { code: 'color-contrast' }, { code: 'other' }, { code: 'color-contrast' } ]
```

then colorContrastInstances will only contain the first color contrast issue? Could you elaborate on why we skip the latter ones? it doesn't seem to match the function name.

KimSarabia replied:
@jdlrobson I found that the first group of color contrast issues are the only ones that are valid. I discovered this while comparing the .csv list of color contrast errors to the list generated by axe-core on the browser.

Typical pattern of pa11y test output:
```
issues [ 
{code: 'color-contrast'}, // <-start: Initial set of color contrast errors - valid error according to axe-core
{code: 'color-contrast'}, // part of initial set - valid
{code: 'color-contrast'}, // <-end: Initial set of color contrast errors - valid
{code: 'aria-attributes-invalid'}, <-start of new set of errors
{code: 'color-contrast'} // <-start of another set of color contrast errors - invalid error according to axe-core
]
```

It is definitely flawed and won't work if it returns a list of **all** invalid errors but the sample output in the PR description is a good sample of the accuracy of the report with this hack. Since there's more time next week, I can try to spend time transitioning to axe and leaving pa11y if we decide this is too unreliable.

KimSarabia replied:
@bernardwang

bernardwang replied:
This seems to work for the case described above, but it assumes that there will always be correct color contrast errors in addition to incorrect ones. if there are only invalid ones, this code will count them all as valid. I'm not sure what's best at this point, maybe this is fine because it gets us closer, but either way its probably fine

Commit:
Add back HTMLReporter

jdlrobson committed:
Merge pull request #7 from wikimedia/reintroduce-html-report

Add back HTMLReporter

Commit:
Merge branch 'main' into filter-out-common-false-flags

bernardwang commented on runA11yTests.js:
```diff
@@ -8,6 +8,9 @@ const util = require( 'util' );
 const htmlReporter = require( path.resolve( __dirname, './reporter/reporter.js' ) );
 const reportTemplate = fs.readFileSync( path.resolve( __dirname, './reporter/report.mustache' ), 'utf8' );
 
+const htmlReporter = require( path.resolve( __dirname, './reporter/reporter.js' ) );
```
this is duplicated, should be removed

Commit:
Remove merge conflict editor error

bernardwang commented on runA11yTests.js:
```diff
@@ -20,6 +20,34 @@ function resetReportDir( config ) {
+			}
+			// Add the current issue to the array
+			colorContrastInstances.push( issue );
+		} else if ( addingToArray && i > 0 && data.issues[i].code !== 'color-contrast' && data.issues[i - 1].code === 'color-contrast' ) {
```
I think we could slightly simplify this by removing addingToArray and instead having this:
```
else if ( colorContrastInstances.length > 0 && data.issues[i].code !== 'color-contrast')
```

addingToArray is never used again after the break, so its not really needed. Same with `data.issues[i - 1].code === 'color-contrast'`, also not really needed i believe

Commit:
Add decorate selector fn

Commit:
Remove unnecessary conditions in extractColorContrastInstances

jdlrobson committed:
Fix post-processing step

The only remaining issue is we need to make sure we call
await page.close();
await browser.close();

at the appropriate moment

jdlrobson committed:
Improve the selector update post-processing to include [style]

jdlrobson committed:
Improve summarize code now we have classes

jdlrobson committed:
Fix multiple use of i attribute, making sure script completes

jdlrobson committed:
Expand elements identified via ID

jdlrobson committed:
Improve summarizing
